src/katgpucbf/xbgpu/bsend.py

```diff
@@ -238,15 +242,17 @@ def send(
         if n_enabled > 0:
             send_futures: list[asyncio.Future] = []
             for frame in self._frames:
-                # TODO (NGC-1232): building this list every time may be too expensive.
-                # Consider caching it and invalidating when streams are enabled/disabled.
-                heaps_to_send = [
-                    spead2.send.HeapReference(heap, substream_index=i, rate=rate)
-                    for i, (heap, enabled) in enumerate(zip(frame.heaps, send_stream.tx_enabled))
-                    if enabled
-                ]
+                if frame.tx_enabled_version != send_stream.tx_enabled_version:
+                    frame.tx_heaps = spead2.send.HeapReferenceList(
+                        [
+                            spead2.send.HeapReference(heap, substream_index=i, rate=rate)
+                            for i, (heap, enabled) in enumerate(zip(frame.heaps, send_stream.tx_enabled))
+                            if enabled
+                        ]
+                    )
```

Comment:
Ah, I see what you mean when you say
> building this list every time may be too expensive.

BSend itself honours the `enable_substream` request and ticks over when it's time to update any substreams. The `Frame`s then follow.

Very cool.